### ISSUE TITLE
Kgdboe does not support CONFIG_KDB_KGDB

### ISSUE THREAD

**Issue by disconnect3d** [sysprogs/kgdboe#15]
Hi,

TL;DR: Kgdboe does not support `CONFIG_KGDB_KGDB`. If this kernel config is enabled, a GDB connection to kgdboe will halt the target system. I haven't exhaustively analyzed this issue, but kdb makes kgdboe gdbstub to return junk kdb shell messages instead of speaking the expected gdbserver protocol.

Full story below.

I have been trying to run kgdboe against Ubuntu 22.04 (5.19.0-38-generic) running on ryzen 5950x cpu and on three different network cards and drivers:
* I211 Gigabit Network Connection with igb driver (firmware 0. 6-1)
* RTL8125 2.5GbE Controller with r8169 driver (firmware 72.daa05125.0)
* A ["Satechi" ethernet/usb adapter](https://satechi.net/products/type-c-2-in-1-usb-3-1-aluminum-3-port-hub-and-ethernet-port?variant=822360473609) with r8152 driver (driverversion=v1.12.13, firmeware=rtl8153a-4 v2 02/07/20)

For this, I build the latest kgdboe from this repo with a `-D__NO_FORTIFY` definition to compile it (to prevent kernel's fortify source mitigation errors, e.g., on `memcpy` calls on whole structs when the struct has padding etc.) as well as I signed the module with my own and enrolled MOK key (I have used [this Ubuntu tutorial for that](https://ubuntu.com/blog/how-to-sign-things-for-secure-boot)). 

Then, all my attempts to run kgdboe ended up with my target system halting even though I used a tested network driver (r8169) and trying out different GDB versions on the host (12.1 on MacOS, 12.0.90 on Ubuntu an self compiled 13.1 on Ubuntu). For completeness, here is how I inserted `kgdboe` to the kernel:
```
sudo insmod kgdboe.ko kallsyms_lookup_name_address=0x$(sudo cat /proc/kallsyms | grep 'T kallsyms_lookup_name' | awk '{print $1}') device_name=enp5s0
```

For what is worth, I also ran the GDB with `set debug remote 1` to investigate what is going on. Here is the full log from such run: https://gist.github.com/disconnect3d/0c4c68ea6ff72830aa581e0a1c9185b1#file-kgdboe-attempt-log

A friend of mine (Thx noodly from irc.gynvael.live) also tested kgdboe on his own Gentoo machine with 6.3.x kernel and it worked for him with a tested (e1000) network driver.

We then compared our kernel configs regarding KGDB. Here is the config from Ubuntu with kernel 5.19.0-38:
```
dc@dc:~$ cat /boot/config-5.19.0-38-generic | grep -i kgdb
CONFIG_SERIAL_KGDB_NMI=y
CONFIG_HAVE_ARCH_KGDB=y
CONFIG_KGDB=y
CONFIG_KGDB_HONOUR_BLOCKLIST=y
CONFIG_KGDB_SERIAL_CONSOLE=y
# CONFIG_KGDB_TESTS is not set
CONFIG_KGDB_LOW_LEVEL_TRAP=y
CONFIG_KGDB_KDB=y
```

And here is the config from Gentoo with 6.3.x kernel - on which kgdboe worked fine:
```
# CONFIG_SERIAL_KGDB_NMI is not set
--
  | CONFIG_HAVE_ARCH_KGDB=y
  | CONFIG_KGDB=y
  | CONFIG_KGDB_HONOUR_BLOCKLIST=y
  | CONFIG_KGDB_SERIAL_CONSOLE=y
  | CONFIG_KGDB_TESTS=y
  | # CONFIG_KGDB_TESTS_ON_BOOT is not set
  | # CONFIG_KGDB_LOW_LEVEL_TRAP is not set
  | # CONFIG_KGDB_KDB is not set
```

Clearly, the configs differ. I went ahead and rebuild an unsigned Ubuntu kernel and changed my config to make it similar to my friend's. Here it is now:
```
dc@dc:~$ cat /boot/config-5.19.0-38-generic-unsigned | grep -i kgdb
CONFIG_SERIAL_KGDB_NMI=y
CONFIG_HAVE_ARCH_KGDB=y
CONFIG_KGDB=y
CONFIG_KGDB_HONOUR_BLOCKLIST=y
CONFIG_KGDB_SERIAL_CONSOLE=y
CONFIG_KGDB_TESTS=y
# CONFIG_KGDB_TESTS_ON_BOOT is not set
# CONFIG_KGDB_LOW_LEVEL_TRAP is not set
# CONFIG_KGDB_KDB is not set
```

On this unsigned kernel, then kgdboe works fine with the r8169 network driver (I haven't really tested others).

However, given all this testing as well as the GDB remote debug output, which said something like:
```
Entering kdb (current=0xfff..., pid 14) on processor 0 due to NonMaskable Interrupt @ 0xfff...
[0] kdb>
diag: 22: Permission Denied
[0] kdb> :multiprocess
```
Along with `[remote] Junk: gqSupporteddiag: -22: Permission denied` and timeouts at the end, I believe the problem results from `kdb` being enabled for kgdb. At least... disabling it as well as `CONFIG_KGDB_LOW_LEVEL_TRAP` worked for me, though, I am not sure if `CONFIG_KGDB_LOW_LEVEL_TRAP` changes anything.

---

With all this, I suggest that kgdboe should error out to compile if KGDB_KDB is enabled. I will send a PR with such a change.

**Comment by sysprogs** [comment]
Thanks for the super-detailed description. We've just rechecked KGDBoE on stock Ubuntu 22.04 (**CONFIG_KGDB_KDB** is enabled) under VMWare (e1000 network driver) and it worked just fine.

The KDB behavior could be caused by an exception somewhere in **kgdb**, forcing the kernel to fall back to the most basic debugging method available.

If you really want to get to the bottom of this, you can try reproducing the problem under VMWare (via PCIe passthrough), running KGDBoE, and attaching another gdb instance to the VMWare gdb stub. This will allow you to debug the kernel even if the debugging facilities of the kernel itself got crashed.

**Comment by disconnect3d** [comment]
> Thanks for the super-detailed description. We've just rechecked KGDBoE on stock Ubuntu 22.04 (CONFIG_KGDB_KDB is enabled) under VMWare (e1000 network driver) and it worked just fine.

Interesting, thanks.

> The KDB behavior could be caused by an exception somewhere in kgdb, forcing the kernel to fall back to the most basic debugging method available.
Hmm... I am wondering: why does it then work on my side if I disable KDB? Can this also be something like `CONFIG_KGDB_LOW_LEVEL_TRAP`? Also, out of curiosity, is this 1 in your VMware setup?:
```
root@dc:/home/dc# cat /sys/module/kdb/parameters/cmd_enable 
1
```
(I don't remember if I tried kgdboe with this parameter set to 0; I doubt it changes anything, but who knows...)

> If you really want to get to the bottom of this, you can try reproducing the problem under VMWare (via PCIe passthrough), running KGDBoE, and attaching another gdb instance to the VMWare gdb stub. This will allow you to debug the kernel even if the debugging facilities of the kernel itself got crashed.
Just to be sure: do you mean to use PCIe passthrough for a network card, or, for something else?

**Comment by sysprogs** [comment]
Our VMWare setup uses the stock kernel provided by Ubuntu 22.04. It does have `CONFIG_KGDB_LOW_LEVEL_TRAP` set.
As for PCIe passthrough, yes, if you could configure VMWare to "move" your network card to the guest, you should be able to test out KGDBoE with it, while being able to debug the entire setup using the VMWare gdb stub.

**Comment by Koston-0xDEADBEEF** [comment]
Probably not so useful information, but I'm seeing the exact same problem, just a rather different platform. It is indeed KGDB throwing exceptions. System doesn't crash, but KDB stops functioning entirely so I can't use kgdboe to debug it.  Might have to resort to caveman methods of just staring at the code until my eyes bleed.

```
Entering kdb (current=0xc1ccb180, pid 551) due to Keyboard Entry
kdb> [   60.992349] ------------[ cut here ]------------
[   60.997025] WARNING: CPU: 0 PID: 551 at net/core/netpoll.c:394 netpoll_send_udp+0xc8/0x450
[   61.005361] Modules linked in: kgdboe(O)
[   61.009330] CPU: 0 PID: 551 Comm: MiSTer Tainted: G        W  O      5.15.1-MiSTer #18
[   61.017258] Hardware name: Altera SOCFPGA
[   61.021309] Backtrace:
[   61.023785] [<c010bf18>] (dump_backtrace) from [<c010c25c>] (show_stack+0x20/0x24)
[   61.031407]  r7:00000009 r6:0000018a r5:c08a1020 r4:c0d63678
[   61.037111] [<c010c23c>] (show_stack) from [<c0582e04>] (dump_stack+0x2c/0x30)
[   61.044343] [<c0582dd8>] (dump_stack) from [<c0124040>] (__warn+0x124/0x178)
[   61.051450]  r5:c08a1020 r4:c0ddd0b4
[   61.055072] [<c0123f1c>] (__warn) from [<c012411c>] (warn_slowpath_fmt+0x88/0xb0)
[   61.062597]  r7:00000009 r6:c08a1020 r5:0000018a r4:c0ddd0b4
[   61.068255] [<c0124098>] (warn_slowpath_fmt) from [<c08a1020>] (netpoll_send_udp+0xc8/0x450)
[   61.076736]  r8:00000001 r7:c10e111c r6:c2190300 r5:000000bb r4:bf00544c
[   61.083454] [<c08a0f58>] (netpoll_send_udp) from [<bf001988>] (netpoll_wrapper_send_reply+0x28/0x34 [kgdboe])
[   61.093457]  r10:c10e4ff0 r9:c10e1118 r8:c1d5bd7c r7:c10e111c r6:000000bb r5:000000bb
[   61.101314]  r4:bf00544c
[   61.103890] [<bf001960>] (netpoll_wrapper_send_reply [kgdboe]) from [<bf0004b8>] (kgdboe_flush+0x44/0x54 [kgdboe])
[   61.114328] [<bf000474>] (kgdboe_flush [kgdboe]) from [<c01aa4f0>] (gdb_serial_stub+0x1180/0x1250)   [   61.123347]  r5:000000bb r4:0000000c
[   61.126932] [<c01a9370>] (gdb_serial_stub) from [<c01a7e5c>] (kgdb_cpu_enter+0x47c/0x690)
[   61.135149]  r10:00000000 r9:ffffcfc7 r8:fffe1dc0 r7:c1d5bd7c r6:c1023c10 r5:c1d5a000
[   61.142984]  r4:c10e1114
[   61.145582] [<c01a79e0>] (kgdb_cpu_enter) from [<c01a867c>] (kgdb_handle_exception+0x118/0x128)
[   61.154323]  r10:00000006 r9:c1d5a000 r8:00000001 r7:00000000 r6:00000000 r5:c1d5be08
[   61.162184]  r4:00000000
[   61.164740] [<c01a8564>] (kgdb_handle_exception) from [<c010ded8>] (kgdb_compiled_brk_fn+0x34/0x3c)
[   61.173827]  r8:60070193 r7:c010dea4 r6:c01a78ec r5:c1d5be08 r4:e7ffdeff                             [   61.180580] [<c010dea4>] (kgdb_compiled_brk_fn) from [<c010c630>] (do_undefinstr+0x184/0x240)
[   61.189133] [<c010c4ac>] (do_undefinstr) from [<c0100b98>] (__und_svc_finish+0x0/0x28)
[   61.197068] Exception stack(0xc1d5be08 to 0xc1d5be50)
[   61.202156] be00:                   bf00500c ffffffff 00000000 00000000 c10e1114 c10e1120
[   61.210368] be20: c100d1ec c10c8bc0 c10c9440 00000040 00000006 c1d5beac c1d5be88 c1d5be98
[   61.218562] be40: c0114090 c01a78ec 60070113 ffffffff
[   61.223646]  r9:c1d5a000 r8:c10c9440 r7:c1d5be3c r6:ffffffff r5:60070113 r4:c01a78f0
[   61.231395] [<c01a78ac>] (kgdb_breakpoint) from [<bf0002f4>] (kgdboe_tasklet_bpt+0x18/0x1c [kgdboe])
[   61.240597]  r5:00000000 r4:bf00500c
[   61.244182] [<bf0002dc>] (kgdboe_tasklet_bpt [kgdboe]) from [<c0128b34>] (tasklet_action_common+0x80/0x100)
[   61.253973] [<c0128ab4>] (tasklet_action_common) from [<c0128bfc>] (tasklet_action+0x20/0x28)
[   61.262523]  r10:c10c9440 r9:00000040 r8:00000006 r7:c1d5a000 r6:00000000 r5:00000007                [   61.270366]  r4:c10d7798 r3:c0128bdc
[   61.273952] [<c0128bdc>] (tasklet_action) from [<c010142c>] (__do_softirq+0x134/0x368)
[   61.281904] [<c01012f8>] (__do_softirq) from [<c012920c>] (irq_exit+0xa4/0xd8)
[   61.289196]  r10:00000000 r9:c108d644 r8:e000210c r7:e0002100 r6:00000000 r5:00000098
[   61.297056]  r4:00000000
[   61.299631] [<c0129168>] (irq_exit) from [<c0164e34>] (handle_domain_irq+0x68/0x84)
[   61.307302] [<c0164dcc>] (handle_domain_irq) from [<c01012b8>] (gic_handle_irq+0x60/0xa0)
[   61.315517]  r7:e0002100 r6:c108d634 r5:c1003944 r4:c1d5bfb0
[   61.321178] [<c0101258>] (gic_handle_irq) from [<c0100e54>] (__irq_usr+0x54/0x80)
[   61.328684] Exception stack(0xc1d5bfb0 to 0xc1d5bff8)
[   61.333767] bfa0:                                     00000000 8000002b 00000000 001f0818
[   61.341954] bfc0: 00000000 001075e0 00000000 001053f0 000f7400 00000000 00000000 00000000
[   61.350146] bfe0: 00000193 b3398078 0005df11 0005df34 40070030 ffffffff
[   61.356785]  r9:00000000 r8:10c5387d r7:10c5387d r6:ffffffff r5:40070030 r4:0005df34
[   61.364568] ---[ end trace 5b2afcc5985b161d ]---
```
